File: src/main/java/io/github/liyze09/arms/network/packet/serverbound/LoginStart.java
package io.github.liyze09.arms.network.packet.serverbound;

import io.github.liyze09.arms.network.Connection;
import io.github.liyze09.arms.network.packet.ServerBoundPacketDecoder;
import io.github.liyze09.arms.network.packet.clientbound.LoginSuccess;
import io.netty.buffer.ByteBuf;
import org.jetbrains.annotations.NotNull;

import static io.github.liyze09.arms.network.PackUtils.*;

public class LoginStart implements ServerBoundPacketDecoder {
    @Override
    public void decode(ByteBuf buf, @NotNull Connection connection) {
        var username = readString(buf);
        var uuid = new Connection.UUID(buf.readLong(), buf.readLong());
        connection.setName(username);
        connection.setUUID(uuid);
        sendPacket(connection,
                new LoginSuccess.LoginSuccessBody(uuid, username),
                LoginSuccess.getInstance()
        );
    }
}

File: src/main/kotlin/io/github/liyze09/arms/network/packet/serverbound/LoginStart.kt
package io.github.liyze09.arms.network.packet.serverbound

import io.github.liyze09.arms.network.Connection
import io.github.liyze09.arms.network.PackUtils.readString
import io.github.liyze09.arms.network.packet.clientbound.LoginSuccess
import io.github.liyze09.arms.network.packet.clientbound.LoginSuccessBody
import io.github.liyze09.arms.network.packet.clientbound.SetCompression
import io.netty.buffer.ByteBuf
import net.minecraftarm.common.UUID

object LoginStart : io.github.liyze09.arms.network.packet.ServerBoundPacketDecoder {
    override fun decode(buf: ByteBuf, connection: Connection) {
        val username = buf.readString(16)
        val uuid = UUID(buf.readLong(), buf.readLong())
        connection.setUsername(username)
        connection.setUUID(uuid)
        connection.sendPacket(
            null,
            SetCompression
        ).sendPacket(
            LoginSuccessBody(uuid, username),
            LoginSuccess
        )
    }
}
